systemtests/upgrade_test.go

```diff
@@ -0,0 +1,113 @@
+//go:build system_test && linux
+
+package systemtests
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"github.com/tidwall/gjson"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+)
+
+func TestChainUpgrade(t *testing.T) {
+	// Scenario:
+	// start a legacy chain with some state
+	// when a chain upgrade proposal is executed
+	// then the chain upgrades successfully
+	sut.StopChain()
+
+	legacyBinary := FetchExecutable(t, "v0.50")
+	t.Logf("+++ legacy binary: %s\n", legacyBinary)
+	currentBranchBinary := sut.execBinary
+	currentInitializer := sut.testnetInitializer
+	sut.SetExecBinary(legacyBinary)
+	sut.SetTestnetInitializer(NewModifyConfigYamlInitializer(legacyBinary, sut))
+	sut.SetupChain()
+	votingPeriod := 5 * time.Second // enough time to vote
+	sut.ModifyGenesisJSON(t, SetGovVotingPeriod(t, votingPeriod))
+
+	const (
+		upgradeHeight int64 = 22
+		upgradeName         = "v050-to-v051"
+	)
+
+	sut.StartChain(t, fmt.Sprintf("--halt-height=%d", upgradeHeight))
+
+	cli := NewCLIWrapper(t, sut, verbose)
+	govAddr := sdk.AccAddress(address.Module("gov")).String()
+	// submit upgrade proposal
+	proposal := fmt.Sprintf(`
+{
+ "messages": [
+  {
+   "@type": "/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade",
+   "authority": %q,
+   "plan": {
+    "name": %q,
+    "height": "%d"
+   }
+  }
+ ],
+ "metadata": "ipfs://CID",
+ "deposit": "100000000stake",
+ "title": "my upgrade",
+ "summary": "testing"
+}`, govAddr, upgradeName, upgradeHeight)
+	proposalID := cli.SubmitAndVoteGovProposal(proposal)
+	t.Logf("current_height: %d\n", sut.currentHeight)
+	raw := cli.CustomQuery("q", "gov", "proposal", proposalID)
+	t.Log(raw)
+
+	sut.AwaitBlockHeight(t, upgradeHeight-1, 60*time.Second)
+	t.Logf("current_height: %d\n", sut.currentHeight)
+	raw = cli.CustomQuery("q", "gov", "proposal", proposalID)
+	proposalStatus := gjson.Get(raw, "proposal.status").String()
+	require.Equal(t, "3", proposalStatus, raw) // PROPOSAL_STATUS_PASSED
+
+	t.Log("waiting for upgrade info")
+	sut.AwaitUpgradeInfo(t)
+	sut.StopChain()
+
+	t.Log("Upgrade height was reached. Upgrading chain")
+	sut.SetExecBinary(currentBranchBinary)
+	sut.SetTestnetInitializer(currentInitializer)
+	sut.StartChain(t)
```

Comment:
here the chain is started with the latest binary (compiled in the Makefile, copied to systemtests/binaries and default in main_test.go)